### PR TITLE
add declarative component definition

### PR DESCRIPTION
**Pull request by flaport**
Hi I added a simple declarative component implementation. It's basically a proof of concept, but I think it could be quite useful.

this is how you could define an mzi:

```python
@gf.declarative_component
class declarative_mzi:
    mmi_in = gf.components.mmi1x2()
    mmi_out = gf.components.mmi2x2()
    straight_top1 = gf.components.straight(length=10.0)
    straight_top2 = gf.components.straight(length=10.0)
    bend_top1 = gf.components.bend_euler()
    bend_top2 = gf.components.bend_euler().mirror()
    bend_top3 = gf.components.bend_euler().mirror()
    bend_top4 = gf.components.bend_euler()
    bend_btm1 = gf.components.bend_euler().mirror()
    bend_btm2 = gf.components.bend_euler()
    bend_btm3 = gf.components.bend_euler()
    bend_btm4 = gf.components.bend_euler().mirror()
    
    def connections(self):
        return [
            (self.bend_top1.ports['o1'], self.mmi_in.ports['o2']),
            (self.straight_top1.ports['o1'], self.bend_top1.ports['o2']),
            (self.bend_top2.ports['o1'], self.straight_top1.ports['o2']),
            (self.bend_top3.ports['o1'], self.bend_top2.ports['o2']),
            (self.straight_top2.ports['o1'], self.bend_top3.ports['o2']),
            (self.bend_top4.ports['o1'], self.straight_top2.ports['o2']),
            (self.bend_btm1.ports['o1'], self.mmi_in.ports['o3']),
            (self.bend_btm2.ports['o1'], self.bend_btm1.ports['o2']),
            (self.bend_btm3.ports['o1'], self.bend_btm2.ports['o2']),
            (self.bend_btm4.ports['o1'], self.bend_btm3.ports['o2']),
            (self.mmi_out.ports["o1"], self.bend_btm4.ports['o2']),
        ]
    
    def ports(self):
        return {
            'o1': self.mmi_in.ports['o1'],
            'o2': self.mmi_out.ports['o3'],
            'o3': self.mmi_out.ports['o4'],
        }
```

As you can see in the PR, it's only a few lines of code to make this work. Let me know what you think. It's just a proof of concept, but if you like it, I think we can extend it to support the full netlist spec.

I also added a notebook which you can try to run to get more information.

### CODE REVIEW

**Comment by flaport** [comment]
Forgot to add the notebook :) Just pushed it.

**Comment by tvt173** [comment]
interesting idea... I like the concise declarative definition, but I get a bit concerned whenever anything is too "magical" or new of a concept to follow... I think it will be hard to follow for a new user i.e. why what they defined as a class got transformed into an object (magically via the decorator)... for example, as an alternative, why not just inherit from a base class which defines a `to_component()` function, or something like that, which is more explicit about what is happening?

overall maybe it would be good to be clear on the objectives of what problem you want to solve here and why this is better than the alternatives

**Comment by flaport** [comment]
Yeah, like I said, just a proof of concept. A to_component() method might indeed be a bit more explicit.

I just quickly coded this up after the developers meeting this morning, where there was a discussion if we should support something more declarative within the python ecosystem. We currently already have the yaml definition, but sometimes staying in python-land gives you a bit more flexibility and of course linting and completion.

**Comment by flaport** [comment]
haha, you merged it already, @joamatab ? I don't think it was fully ready or agreed upon, but I guess we can change/update it later.

**Comment by tvt173** [comment]
sure, I understand the high-level motivation. but to me, a class definition that magically turns into an object violates the [Principle of Least Astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) 😄 . i don't think it will be straightforward for most users, especially when layered with being defined inside a function for a pcell, then wrapped with a cell decorator, as in your notebook example.. it gets very convoluted very quickly. i think we could still improve it

**Comment by flaport** [comment]
Yeah, I agree. Like I said it was something I coded up in less than half an hour. I'm pretty sure there are plenty of changes that would make it better and less magical. I was pretty happy to see it was very easy to implement something like this though (although the code I submitted is admittedly pretty hacky)

**Comment by flaport** [comment]
To reduce the astonishment factor, you could also use it as follows:

```python
@gf.cell
def my_component():
    class ComponentDeclaration:
        ...
        def connections(self):
            ...
        def ports(self):
           ...
    return gf.declarative_component(ComponentDeclaration)
```

Although, I admit... Defining a class inside the body of a function still has some weirdness to it 😅

**Comment by flaport** [comment]
Here is an alternative implementation (subclass DeclarativeCell in stead of the usual gf.cell decorator on a function to get a PCell):

```python
class mzi(DeclarativeCell):
    delta_length: float = 10.0
    
    def instances(self):
        self.mmi_in = gf.components.mmi1x2()
        self.mmi_out = gf.components.mmi2x2()
        self.straight_top1 = gf.components.straight(length=self.delta_length/2)
        self.straight_top2 = gf.components.straight(length=self.delta_length/2)
        self.bend_top1 = gf.components.bend_euler()
        self.bend_top2 = gf.components.bend_euler().mirror()
        self.bend_top3 = gf.components.bend_euler().mirror()
        self.bend_top4 = gf.components.bend_euler()
        self.bend_btm1 = gf.components.bend_euler().mirror()
        self.bend_btm2 = gf.components.bend_euler()
        self.bend_btm3 = gf.components.bend_euler()
        self.bend_btm4 = gf.components.bend_euler().mirror()
        
    def connections(self):
        return [
            (self.bend_top1.ports['o1'], self.mmi_in.ports['o2']),
            (self.straight_top1.ports['o1'], self.bend_top1.ports['o2']),
            (self.bend_top2.ports['o1'], self.straight_top1.ports['o2']),
            (self.bend_top3.ports['o1'], self.bend_top2.ports['o2']),
            (self.straight_top2.ports['o1'], self.bend_top3.ports['o2']),
            (self.bend_top4.ports['o1'], self.straight_top2.ports['o2']),
            (self.bend_btm1.ports['o1'], self.mmi_in.ports['o3']),
            (self.bend_btm2.ports['o1'], self.bend_btm1.ports['o2']),
            (self.bend_btm3.ports['o1'], self.bend_btm2.ports['o2']),
            (self.bend_btm4.ports['o1'], self.bend_btm3.ports['o2']),
            (self.mmi_out.ports["o1"], self.bend_btm4.ports['o2']),
        ]
    
    def ports(self):
        return {
            'o1': self.mmi_in.ports['o1'],
            'o2': self.mmi_out.ports['o3'],
            'o3': self.mmi_out.ports['o4'],
        }
```

Here's the scary 18 lines of code that make this work (I know... terribly hacky for now xD)

```python
from pydantic import BaseModel
class DeclarativeCell(gf.Component, BaseModel):
    def __init__(self, *args, **kwargs):
        BaseModel.__init__(self)
        ports_func = self.ports
        gf.Component.__init__(self)
        self.instances()
        for k, v in self.__dict__.items():
            if not isinstance(v, gf.Component):
                continue
            ref = (self << v)
            setattr(self, k, ref)
        for p1, p2 in self.connections():
            p1.reference.connect(p1.name, p2.reference.ports[p2.name])
        for name, p in ports_func().items():
            self.add_port(name, port=p.reference.ports[p.name])
    class Config:
        extra = "allow"
```

**Comment by flaport** [comment]
This is my favorite version, as opposed to the previous two versions, because:
- compared to 1: you get a PCell function by default in stead of a gf.Component, which is at least less weird.
- compared to 2: calling the resulting PCell function gets you a normal gf.Component (in stead of a weird subclass combination of a gf.Component and a pydantic.BaseModel).

Here's the implementation:

```python
import gdsfactory as gf
from dataclasses import dataclass
from functools import wraps
def declarative_cell(cls):
    cls = dataclass(cls)
    @wraps(cls)
    def cell(*args, **kwargs):
        decl = cls(*args, **kwargs)
        decl.instances()
        comp = gf.Component()
        for k, c in vars(decl).items():
            if not isinstance(c, gf.Component):
                continue
            ref = (comp << c)
            setattr(comp, k, ref)
            setattr(decl, k, ref)
        for p1, p2 in decl.connections():
            p1.reference.connect(p1.name, p2.reference.ports[p2.name])
        for name, p in decl.ports().items():
            comp.add_port(name, port=p.reference.ports[p.name])
        return comp
    return cell
```

Here is an example usage:

```python
@declarative_cell
class mzi:
    delta_length: float = 3.0
    
    def instances(self):
        self.mmi_in = gf.components.mmi1x2()
        self.mmi_out = gf.components.mmi2x2()
        self.straight_top1 = gf.components.straight(length=self.delta_length/2)
        self.straight_top2 = gf.components.straight(length=self.delta_length/2)
        self.bend_top1 = gf.components.bend_euler()
        self.bend_top2 = gf.components.bend_euler().mirror()
        self.bend_top3 = gf.components.bend_euler().mirror()
        self.bend_top4 = gf.components.bend_euler()
        self.bend_btm1 = gf.components.bend_euler().mirror()
        self.bend_btm2 = gf.components.bend_euler()
        self.bend_btm3 = gf.components.bend_euler()
        self.bend_btm4 = gf.components.bend_euler().mirror()
        
    def connections(self):
        return [
            (self.bend_top1.ports['o1'], self.mmi_in.ports['o2']),
            (self.straight_top1.ports['o1'], self.bend_top1.ports['o2']),
            (self.bend_top2.ports['o1'], self.straight_top1.ports['o2']),
            (self.bend_top3.ports['o1'], self.bend_top2.ports['o2']),
            (self.straight_top2.ports['o1'], self.bend_top3.ports['o2']),
            (self.bend_top4.ports['o1'], self.straight_top2.ports['o2']),
            (self.bend_btm1.ports['o1'], self.mmi_in.ports['o3']),
            (self.bend_btm2.ports['o1'], self.bend_btm1.ports['o2']),
            (self.bend_btm3.ports['o1'], self.bend_btm2.ports['o2']),
            (self.bend_btm4.ports['o1'], self.bend_btm3.ports['o2']),
            (self.mmi_out.ports["o1"], self.bend_btm4.ports['o2']),
        ]
    
    def ports(self):
        return {
            'o1': self.mmi_in.ports['o1'],
            'o2': self.mmi_out.ports['o3'],
            'o3': self.mmi_out.ports['o4'],
        }
```

Note that the example looks a lot like the first example, but you get a PCell function rather than an object. I know... you still get a function rather than a class which might still annoy Troy a bit, but the concept is cleaner: when you call a class you expect an object. When you call the result of this definition you also get an object: less weird for sure.

**Comment by joamatab** [comment]
Loooks great, thank you Floris, shall we add the notebook you made to the docs?

**Comment by flaport** [comment]
sure... Feel free to add. I added it to the docs folder, but I'm a bit unsure how to do the rest :)

**Comment by tvt173** [comment]
ok, if you insist @flaport , I'll continue to pick on this... 😋 

it seems inconsistent and eventually troublesome to me that both `connections()` and `ports()` are methods that return the respective values, but `instances()` does not... (it instead adds instances directly as attributes to the object)

so here are my issues with this...
1. It's confusing... by name/style, it has exactly the same form as the other two methods, but it has different behavior.
2. It sets the instances directly to the main object, which I assume is meant for convenience accessing those instances later, but could cause real confusion and significant limitations... For example, the way you have it coded now, you cannot have a parameter which is a component, because it will be interpreted as an instance... This is a real scenario, i.e. if you want to set the taper to be used in an MMI as a parameter (then instance it n times at each input/output).

I would suggest either returning `instances()` as a dictionary of instances, or as a sub-object, if you want to keep it strongly typed. That way the behavior will be more consistent with the other methods, and you can simplify the logic in `declarative_cell`, as it will be very clear now how to get instances (it won't be confusable with parameters to the class)

**Comment by joamatab** [comment]
yes, i think instances should return references

I also had a hard time understanding the magic of some of the code :)
